contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,508 @@
+                sourceAddress: source_address.to_alloc_vec().into(),
+                destinationAddress: destination_address.to_alloc_vec().into(),
+                amount: amount.try_into().expect("failed to convert"),
+                data: into_vec(data).into(),
+            }
+            .abi_encode_params(),
+            Self::DeployInterchainToken(types::DeployInterchainToken {
+                token_id,
+                name,
+                symbol,
+                decimals,
+                minter,
+            }) => DeployInterchainToken {
+                messageType: MessageType::DeployInterchainToken.into(),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                name: to_std_string(name),
+                symbol: to_std_string(symbol),
+                decimals,
+                minter: into_vec(minter).into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let payload = payload.to_alloc_vec();
+
+        let message_type = MessageType::abi_decode(&payload[0..32], true)
+            .map_err(|_| MessageError::InvalidMessageType)?;
+
+        let message = match message_type {
+            MessageType::InterchainTransfer => {
+                let decoded = InterchainTransfer::abi_decode_params(&payload, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::InterchainTransfer(types::InterchainTransfer {
+                    token_id: BytesN::from_array(env, &decoded.tokenId.into()),
+                    source_address: Bytes::from_slice(env, decoded.sourceAddress.as_ref()),
+                    destination_address: Bytes::from_slice(
+                        env,
+                        decoded.destinationAddress.as_ref(),
+                    ),
+                    amount: to_i128(decoded.amount)?,
+                    data: from_vec(env, decoded.data.as_ref()),
+                })
+            }
+            MessageType::DeployInterchainToken => {
+                let decoded = DeployInterchainToken::abi_decode_params(&payload, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::DeployInterchainToken(types::DeployInterchainToken {
+                    token_id: BytesN::from_array(env, &decoded.tokenId.into()),
+                    name: String::from_str(env, &decoded.name),
+                    symbol: String::from_str(env, &decoded.symbol),
+                    decimals: decoded.decimals,
+                    minter: from_vec(env, decoded.minter.as_ref()),
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+#[allow(dead_code)]
```

Comment:
The compiler complains about unused code since these have not been implemented in the ITS contract yet. I have created a ticket to remove the `#[allow(dead_code)]` once the ITS contract is actually using the HubMessage encode/decode methods.